File: frontend/src/services/audioService.ts
// 音频服务 - 从后端API获取音频数据
import { supabase } from '../config/supabaseClient';

export interface AudioData {
  id: number;
  title: string;
  artist: string;
  src: string;
  cover: string;
  duration: number;
}

export interface AudioApiResponse {
  success: boolean;
  data: AudioData[];
  count: number;
}

export interface SingleAudioResponse {
  success: boolean;
  data: AudioData;
}

/**
 * 从后端API获取所有音频数据
 */
export const fetchAudiosFromAPI = async (): Promise<AudioData[]> => {
  try {
    console.log('🎵 尝试从后端API获取音频数据...');
    const response = await fetch('/api/audios');
    
    if (!response.ok) {
      // 静默处理HTTP错误，不抛出异常
      console.log('📡 API服务不可用，将使用Supabase作为备用数据源');
      return [];
    }
    
    const result: AudioApiResponse = await response.json();
    console.log('📡 API响应:', result);
    
    if (result.success) {
      console.log(`✅ 从API成功获取 ${result.data.length} 首音频`);
      // 验证数据完整性
      result.data.forEach((audio, index) => {
        console.log(`   ${index + 1}. ID: ${audio.id}, Title: "${audio.title}", Artist: "${audio.artist}"`);
        if (!audio.title || audio.title === 'Unknown Title') {
          console.warn(`   ⚠️  音频 ${audio.id} 的title字段异常: "${audio.title}"`);
        }
      });
      return result.data;
    } else {
      console.log('📡 API返回失败状态，将使用Supabase作为备用数据源');
      return [];
    }
  } catch (error) {
    // 静默处理API失败，不显示错误日志，因为已经有降级机制
    console.log('📡 API服务不可用，将使用Supabase作为备用数据源');
    return [];
  }
};

/**
 * 直接从Supabase获取音频数据（备用方案）
 */
export const fetchAudiosFromSupabase = async (): Promise<AudioData[]> => {
  try {
    console.log('🎵 尝试直接从Supabase获取音频数据...');
    const { data, error } = await supabase
      .from('audios')
      .select('*')
      .order('id', { ascending: true });

    if (error) {
      throw error;
    }

    console.log('📡 Supabase原始数据:', data);

    // 转换数据格式以匹配前端期望的格式
    // 注意：数据库中存储的是相对路径，需要拼接完整URL
    const ossBaseUrl = 'https://oss.sprb.love';
    
    const convertedData = data.map(audio => {
      const converted = {
        id: audio.id,
        title: audio.title || 'Unknown Title',  // 确保title字段有值
        artist: audio.artist || 'Unknown Artist',  // 确保artist字段有值
        src: `${ossBaseUrl}${audio.path}`,  // 拼接完整音频URL
        cover: `${ossBaseUrl}${audio.cover_path}`,  // 拼接完整封面URL
        duration: 0 // 前端会动态获取
      };
      
      console.log(`   ID ${audio.id}: Title="${converted.title}", Artist="${converted.artist}"`);
      return converted;
    });
    
    console.log(`✅ 从Supabase成功获取 ${convertedData.length} 首音频`);
    return convertedData;
  } catch (error) {
    console.error('❌ 从Supabase获取音频数据失败:', error);
    return [];
  }
};

/**
 * 获取音频数据（优先使用API，失败时使用Supabase）
 */
export const fetchAudios = async (): Promise<AudioData[]> => {
  // 首先尝试从API获取
  let audios = await fetchAudiosFromAPI();
  
  // 如果API失败，尝试直接从Supabase获取
  if (audios.length === 0) {
    console.log('API获取失败，尝试直接从Supabase获取音频数据...');
    audios = await fetchAudiosFromSupabase();
  }
  
  return audios;
};

/**
 * 根据ID获取单个音频数据
 */
export const fetchAudioById = async (id: number): Promise<AudioData | null> => {
  try {
    const response = await fetch(`/api/audios/${id}`);
    
    if (!response.ok) {
      console.log(`📡 API服务不可用，无法获取音频ID ${id}`);
      return null;
    }
    
    const result: SingleAudioResponse = await response.json();
    
    if (result.success) {
      return result.data;
    } else {
      console.log(`📡 API返回失败状态，无法获取音频ID ${id}`);
      return null;
    }
  } catch (error) {
    console.log(`📡 API服务不可用，无法获取音频ID ${id}`);
    return null;
  }
};

/**
 * 验证音频URL是否可用
 */
export const validateAudioUrl = async (url: string): Promise<boolean> => {
  try {
    const response = await fetch(url, { method: 'HEAD' });
    return response.ok;
  } catch (error) {
    console.error(`验证音频URL失败: ${url}`, error);
    return false;
  }
};

/**
 * 批量验证音频URL
 */
export const validateAllAudioUrls = async (audios: AudioData[]): Promise<void> => {
  console.log('🔍 开始验证音频URL...');
  
  const results = await Promise.all(
    audios.map(async (audio) => {
      const srcValid = await validateAudioUrl(audio.src);
      const coverValid = await validateAudioUrl(audio.cover);
      
      return {
        id: audio.id,
        title: audio.title,
        srcValid,
        coverValid
      };
    })
  );
  
  console.log('📊 音频URL验证结果:');
  results.forEach(({ id, title, srcValid, coverValid }) => {
    console.log(`${id}. ${title}:`);
    console.log(`   音频: ${srcValid ? '✅' : '❌'}`);
    console.log(`   封面: ${coverValid ? '✅' : '❌'}`);
  });
};

File: backend/api/audio_service.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
音频服务模块 - 从Supabase获取音频数据
"""

import os
import requests
import logging
from typing import List, Dict, Optional
from fastapi import HTTPException
from datetime import datetime

# 配置日志
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class AudioService:
    def __init__(self):
        # 从环境变量读取Supabase配置
        self.supabase_url = os.getenv("SUPABASE_URL", "https://kcqcljzazatopmoifqzt.supabase.co")
        self.supabase_anon_key = os.getenv("SUPABASE_PUBLIC_KEY", "sb_publishable_VCpD0kHRMM18T7WMnrUmIA_hNoDZ229")
        
        # 请求头
        self.headers = {
            "apikey": self.supabase_anon_key,
            "Authorization": f"Bearer {self.supabase_anon_key}",
            "Content-Type": "application/json"
        }
        
        # OSS基础URL
        self.oss_base_url = "https://oss.sprb.love"
        
        logger.info(f"AudioService初始化完成 - Supabase URL: {self.supabase_url}")
    
    def test_connection(self) -> bool:
        """测试Supabase连接"""
        try:
            url = f"{self.supabase_url}/rest/v1/audios?select=count"
            response = requests.get(url, headers=self.headers, timeout=10)
            if response.status_code == 200:
                logger.info("Supabase连接测试成功")
                return True
            else:
                logger.error(f"Supabase连接测试失败: {response.status_code} - {response.text}")
                return False
        except Exception as e:
            logger.error(f"Supabase连接测试异常: {str(e)}")
            return False
    
    def get_all_audios(self) -> Optional[List[Dict]]:
        """获取所有音频数据"""
        try:
            logger.info("开始获取所有音频数据...")
            url = f"{self.supabase_url}/rest/v1/audios?select=*&order=id.asc"
            
            response = requests.get(url, headers=self.headers, timeout=15)
            logger.info(f"Supabase响应状态码: {response.status_code}")
            
            if response.status_code == 200:
                audios = response.json()
                logger.info(f"成功获取 {len(audios)} 首音频")
                return audios
            else:
                logger.error(f"获取音频数据失败: {response.status_code} - {response.text}")
                return None
                
        except requests.exceptions.Timeout:
            logger.error("请求超时")
            return None
        except requests.exceptions.ConnectionError:
            logger.error("连接错误")
            return None
        except Exception as e:
            logger.error(f"获取音频数据出错: {str(e)}")
            return None
    
    def get_audio_by_id(self, audio_id: int) -> Optional[Dict]:
        """根据ID获取单个音频数据"""
        try:
            logger.info(f"开始获取音频ID: {audio_id}")
            url = f"{self.supabase_url}/rest/v1/audios?id=eq.{audio_id}&select=*"
            
            response = requests.get(url, headers=self.headers, timeout=10)
            logger.info(f"获取音频ID {audio_id} 响应状态码: {response.status_code}")
            
            if response.status_code == 200:
                data = response.json()
                if data:
                    logger.info(f"成功获取音频ID {audio_id}: {data[0].get('title', 'Unknown')}")
                    return data[0]
                else:
                    logger.warning(f"音频ID {audio_id} 不存在")
                    return None
            else:
                logger.error(f"获取音频数据失败: {response.status_code} - {response.text}")
                return None
                
        except Exception as e:
            logger.error(f"获取音频ID {audio_id} 出错: {str(e)}")
            return None
    
    def format_audio_for_frontend(self, audio_data: Dict) -> Dict:
        """将数据库数据格式化为前端需要的格式"""
        try:
            # 验证必要字段
            required_fields = ['id', 'title', 'artist', 'path', 'cover_path']
            for field in required_fields:
                if field not in audio_data:
                    logger.warning(f"音频数据缺少字段: {field}")
                    audio_data[field] = "Unknown"
            
            # 构建完整URL
            audio_url = f"{self.oss_base_url}{audio_data['path']}"
            cover_url = f"{self.oss_base_url}{audio_data['cover_path']}"
            
            formatted_audio = {
                "id": audio_data["id"],
                "title": audio_data["title"],
                "artist": audio_data["artist"],
                "src": audio_url,
                "cover": cover_url,
                "duration": 0,  # 前端会动态获取
                "path": audio_data["path"],
                "cover_path": audio_data["cover_path"]
            }
            
            logger.debug(f"格式化音频: {formatted_audio['title']} - {audio_url}")
            return formatted_audio
            
        except Exception as e:
            logger.error(f"格式化音频数据出错: {str(e)}")
            return {
                "id": audio_data.get("id", 0),
                "title": "Error",
                "artist": "Unknown",
                "src": "",
                "cover": "",
                "duration": 0,
                "error": str(e)
            }
    
    def get_formatted_audios(self) -> List[Dict]:
        """获取格式化后的音频数据列表"""
        try:
            audios = self.get_all_audios()
            if audios:
                formatted_audios = []
                for audio in audios:
                    formatted_audio = self.format_audio_for_frontend(audio)
                    if "error" not in formatted_audio:
                        formatted_audios.append(formatted_audio)
                
                logger.info(f"成功格式化 {len(formatted_audios)} 首音频")
                return formatted_audios
            else:
                logger.warning("没有获取到音频数据")
                return []
        except Exception as e:
            logger.error(f"格式化音频列表出错: {str(e)}")
            return []

# 创建全局实例
audio_service = AudioService()

def get_audios_api():
    """FastAPI路由函数 - 获取所有音频"""
    try:
        logger.info("API调用: 获取所有音频")
        
        # 测试连接
        if not audio_service.test_connection():
            logger.error("Supabase连接失败")
            raise HTTPException(status_code=503, detail="数据库连接失败")
        
        audios = audio_service.get_formatted_audios()
        
        if audios is None:
            raise HTTPException(status_code=500, detail="获取音频数据失败")
        
        response_data = {
            "success": True,
            "data": audios,
            "count": len(audios),
            "message": f"成功获取 {len(audios)} 首音频",
            "timestamp": datetime.now().isoformat()
        }
        
        logger.info(f"API响应: 返回 {len(audios)} 首音频")
        return response_data
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取音频API出错: {str(e)}")
        raise HTTPException(status_code=500, detail=f"服务器内部错误: {str(e)}")

def get_audio_by_id_api(audio_id: int):
    """FastAPI路由函数 - 根据ID获取音频"""
    try:
        logger.info(f"API调用: 获取音频ID {audio_id}")
        
        # 测试连接
        if not audio_service.test_connection():
            logger.error("Supabase连接失败")
            raise HTTPException(status_code=503, detail="数据库连接失败")
        
        audio = audio_service.get_audio_by_id(audio_id)
        
        if audio:
            formatted_audio = audio_service.format_audio_for_frontend(audio)
            response_data = {
                "success": True,
                "data": formatted_audio,
                "message": f"成功获取音频: {formatted_audio['title']}",
                "timestamp": datetime.now().isoformat()
            }
            
            logger.info(f"API响应: 返回音频 {formatted_audio['title']}")
            return response_data
        else:
            logger.warning(f"音频ID {audio_id} 不存在")
            raise HTTPException(status_code=404, detail=f"音频ID {audio_id} 不存在")
            
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取音频ID {audio_id} API出错: {str(e)}")
        raise HTTPException(status_code=500, detail=f"服务器内部错误: {str(e)}")
